### ISSUE TITLE
Cannot depend on ```sckit-sparse``` in a clean environment due to early ```numpy``` import.

### ISSUE THREAD

**Issue by rgiordan** [scikit-sparse/scikit-sparse#55]
I'd like to depend on ```scikit-sparse```, but [this line](https://github.com/scikit-sparse/scikit-sparse/blob/e35b764ae3910615322ac9db64330cee426d05e8/setup.py#L34) in ```setup.py``` imports ```numpy``` before it is installed with ```setup_requires```.   As a consequence, ```scikit-sparse``` (and packages that depend on it) can't be ```pip``` installed in clean virtual environments like Travis.

Of course, one can make a workaround by first installing ```numpy```, but I think the issue would be better fixed in ```setup.py```, e.g. with [something like this](https://stackoverflow.com/questions/19919905/how-to-bootstrap-numpy-installation-in-setup-py).

**Comment by njsmith** [comment]
That answer's obsolete. And setup_requires is unfortunately broken and
should never be used. The modern way to do this is to declare the
dependency in pyproject.toml, and then pip will make sure it's installed
before invoking setup.py.

**Comment by anntzer** [comment]
... except for the fact that support pyproject.toml has at least one longstanding bug (https://github.com/pypa/pip/issues/5317).

I don't actually manage this package at all anymore so my opinion hardly matters, but I do still prefer the build_ext-subclassing solution.

**Comment by njsmith** [comment]
setup_requires has many more outstanding bugs, and they're less likely to
be fixed.

**Comment by anntzer** [comment]
Oh, I have no doubt that pyproject.toml is the way of the future given that the powers to be have decided that.  All I can see, however, is that as of today, setup_requires works for me and pyproject.toml doesn't.

For example, matplotlib has a setup_requires on numpy.  We get many complaints about various aspects of our build, but they are mostly related to the system dependencies (freetype, libpng); I don't remember seeing a single setup_requires-related bug report in the past few years.

**Comment by njsmith** [comment]
The main two issues are that it can't install from wheels, so your users probably end up with super-slow numpy, and it corrupts the environment metadata, so later upgrades/uninstalls don't work right. I think the reasons matplotlib doesn't get bug reports are because (a) most of your users are installing from wheels so they never run setup.py anyway, (b) since the effect is to break numpy, possibly months later, the affected users don't realize that they should complain to matplotlib.

Of course a lot of the blame here rests with the packaging infrastructure for having a footgun like this, and with the companies who won't pay for maintenance of the infrastructure they rely on, etc. But here and now, avoiding `setup_requires` is still a good idea.

**Comment by anntzer** [comment]
OK, let's try to install the following setup.py in a clean, new venv:
```
from setuptools import setup
from setuptools.command.install import install

class install(install):
    def run(self):
        import numpy; print(numpy, file=open("/tmp/log", "w"))
        return super().run()

setup(
    name="foobarbaz",
    cmdclass={"install": install},
    setup_requires=["numpy==1.16.4"],
)
```
One can inspect /tmp/log and check that numpy was indeed available during the install, but after the install, it is not there anymore.

One can even pre-install another version of numpy before running `pip install .`, and check that during the install, the numpy version is 1.16.4 but after the install, we're back to the previously installed version.  Given that the temporarily-setup_installed-numpy is actually installed as an egg, I just don't see how the actually outside-of-setup_install installed numpy is affected.  [Obviously, if we're talking about egg-less setup dependencies, then that won't work -- but then one should ask why they are egg-less.]

Also that means that the sole point at which one has an egg-installed numpy rather than a wheel is during the install, where I don't think the performance of numpy matters that much.

Yes, it is true that setuptools provides a lot of ways to shoot oneself in the foot.  However, that does not mean it is not powerful enough to write correct setup scripts.

**Comment by njsmith** [comment]
Huh. That doesn't match the received wisdom I'd picked up from e.g. the setuptools maintainers, and is much better than I thought. I would still be wary of being bitten by edge cases, and I'm still dubious about the idea of intentionally choosing the older, more complicated, deprecated approach over the newer, simpler, actively-maintained approach, but it does handle the worst issues I know of.

**Comment by anntzer** [comment]
I think a lot of the discussions surrounding setup_requires confused two issues.

Yes, the API around setup_requires (basically, requiring to subclass build_ext for most practical purposes) is less than optimal.  And yes I am aware that because setup_requires relies on easy_install, which doesn't handle wheels, may be less customizable than pip, yada yada -- although I don't know if there are any technical reasons why setup_requires (possibly under a new name) could not have been switched to use pip instead.

On the other hand, this does not mean that setup_requires is underpowered and that a whole new system had to be invented.  And again, that system actually doesn't work in combination with other well-documented configuration options (namely, defaulting to --user installs -- note that this is something that has *also* been repeatedly proposed on distutils-sig, not some obscure edge case).

**Comment by njsmith** [comment]
> I don't know if there are any technical reasons why setup_requires (possibly under a new name) could not have been switched to use pip instead.

There were challenges there, mostly that the only way pip could figure out what needed to be installed was to invoke setup.py, and invoking setup.py already triggered the install of setup_requires. But with modern pip + modern setuptools, setup_requires are basically converted into pyproject.toml requirements and handled by pip: https://github.com/pypa/setuptools/blob/d36295aae2d6d2238546309a0dc4043f6e27e78a/setuptools/build_meta.py#L144

But this isn't really an argument for using setup_requires. In any situation where pip can handle setup_requires, it can also handle pyproject.toml (they use the same code paths internally), and pyproject.toml is significantly simpler to use. The only reason to favor setup_requires is if there's some situation where you want to get legacy easy_install behavior.

> On the other hand, this does not mean that setup_requires is underpowered and that a whole new system had to be invented.

I mean... the user experience around setup_requires is terrible, as you noted, and that's completely unfixable without a new system. Also, this is part of a larger project to break the tight coupling between python packaging and setuptools, which also requires a new system.

>  And again, that system actually doesn't work in combination with other well-documented configuration options (namely, defaulting to --user installs -- note that this is something that has also been repeatedly proposed on distutils-sig, not some obscure edge case).

Yes, software has bugs, I'm not sure what you want me to say there. I get its frustrating when your personal setup breaks. I'm sure they'd welcome help fixing them. And unfortunately using pip.conf to default to --user installs is, in fact, a pretty obscure edge case. Maybe it won't be someday, that would be great, but right now the reason it's not the default yet is because no-one's done the work to go through and figure out how to smooth out the edge cases. And then combine that with editable installs, which are another huge pile of technical debt that we're trying to gradually crawl out from under... it's tough.

**Comment by anntzer** [comment]
>> I don't know if there are any technical reasons why setup_requires (possibly under a new name) could not have been switched to use pip instead.

> There were challenges there, mostly that the only way pip could figure out what needed to be installed was to invoke setup.py, and invoking setup.py already triggered the install of setup_requires. But with modern pip + modern setuptools, setup_requires are basically converted into pyproject.toml requirements and handled by pip: pypa/setuptools:setuptools/build_meta.py@d36295a#L144 [...]

It's great that handling for setup_requires has been improved, I'm all for it.  Whether that had to be done at the same time as introducing a new API is the question.

>> On the other hand, this does not mean that setup_requires is underpowered and that a whole new system had to be invented.

> I mean... the user experience around setup_requires is terrible, as you noted, and that's completely unfixable without a new system.

A pretty obvious solution would have been to make `ext_modules` accept a callable that returns a list of extensions (and then you can import whatever you require from within that callable).  Clearly that solution has been discovered again and again, see e.g. https://github.com/pypa/setuptools_scm/#importing-in-setuppy for something similar.

> Also, this is part of a larger project to break the tight coupling between python packaging and setuptools, which also requires a new system.

Yeah, that would be another great debate to have.  But perhaps another time?  :)

>> And again, that system actually doesn't work in combination with other well-documented configuration options (namely, defaulting to --user installs -- note that this is something that has also been repeatedly proposed on distutils-sig, not some obscure edge case).

> Yes, software has bugs, I'm not sure what you want me to say there. I get its frustrating when your personal setup breaks. I'm sure they'd welcome help fixing them.

Oh, I certainly agree with the fact that at the end of the day, the one who writes the code is the one who gets a say.  But my hands are already full enough maintaining matplotlib that I'll just look at this from the peanut gallery.

> And unfortunately using pip.conf to default to --user installs is, in fact, a pretty obscure edge case. Maybe it won't be someday, that would be great, but right now the reason it's not the default yet is because no-one's done the work to go through and figure out how to smooth out the edge cases. And then combine that with editable installs, which are another huge pile of technical debt that we're trying to gradually crawl out from under... it's tough.

Well, people have been talking about defaulting to --user for more than 5 years now (https://github.com/pypa/pip/issues/1668), and in fact setting that default in pip.conf has worked for me for years (and still does) -- whether on system pythons or on conda pythons -- with no problem.  I don't really want to re-read the whole pip thread to resummarize it here properly but IIRC the main issues were related to policy (how to interact well with the distros?  should we make sure ~/.local/bin is in the $PATH?  etc.) and not to technical issues -- and I was aware of none until pyproject.toml showed up.

**Comment by jellis18** [comment]
@njsmith, @anntzer, @rgiordan There is a PR to fix this in #64 but I don't think this project is maintained anymore. Do any of you have privileges for this project?

I've also re-factored further in a [fork](https://github.com/jellis18/sksparse) and was just going to make releases base on that but I wanted to try to do things through this project.

@jor- are you still out there?

**Comment by anntzer** [comment]
I'm happy to give you commit rights & PyPI ownership, as this project is basically unmaintained now AFAICT.  Let's just give a few days for the other devs to chime in, in case they want to comment on that.

**Comment by jellis18** [comment]
@anntzer thank you for that!

**Comment by jor-** [comment]
Yes, I am still alive. But right now I do not have time to maintain this package anymore. So if there is someone able and willing to maintin this package I am very happy. I hope to have more time for this package again sometime in the future.

**Comment by anntzer** [comment]
Gave you access on PyPI.  I don't manage the conda-forge package so you'll have to figure out the contact point yourself...

**Comment by jellis18** [comment]
@anntzer, any chance I can have admin access so I can set up some secrets to use in github actions?

**Comment by anntzer** [comment]
It's all yours.

**Comment by jellis18** [comment]
Fixed in master and v0.4.5 and is available on pypi and conda-forge.
